lib/posix-process/signal/clone.c

```diff
@@ -0,0 +1,158 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <string.h>
+
+#include <uk/config.h>
+#include <uk/arch/lcpu.h>
+#include <uk/process.h>
+#include <uk/print.h>
+
+#include "signal.h"
+
+static int uk_posix_clone_sighand(const struct clone_args *cl_args,
+				  size_t cl_args_len __unused,
+				  struct uk_thread *child __unused,
+				  struct uk_thread *parent __unused)
+{
+	struct posix_process *pp; /* parent process */
+	struct posix_process *cp; /* child process */
+	struct posix_thread *ct; /* child thread */
+	pid_t ppid; /* parent pid */
+	pid_t cpid; /* child pid */
+	int rc;
+
+	ppid = ukthread2pid(parent);
+	cpid = ukthread2pid(child);
+
+	pp = pid2pprocess(ppid);
+	cp = pid2pprocess(cpid);
+
+	ct = tid2pthread(ukthread2tid(child));
+
+	/* CLONE_SIGHAND and CLONE_CLEAR_SIGHAND should not be together */
+	if (unlikely((cl_args->flags & (CLONE_SIGHAND | CLONE_CLEAR_SIGHAND))
+		     == (CLONE_SIGHAND | CLONE_CLEAR_SIGHAND)))
+		return -EINVAL;
+
+	/* CLONE_SIGHAND requires CLONE_VM */
+	if (unlikely((cl_args->flags & CLONE_SIGHAND) &&
+		     !(cl_args->flags & CLONE_VM)))
+		return -EINVAL;
+
+	/* CLONE_THREAD requires CLONE_SIGHAND */
+	if (unlikely((cl_args->flags & CLONE_THREAD) &&
+		     !(cl_args->flags & CLONE_SIGHAND)))
+		return -EINVAL;
+
+	/* Initialize the new thread's signal descriptor */
+	rc = pprocess_signal_tdesc_alloc(ct);
+	if (unlikely(rc)) {
+		uk_pr_err("Could not allocate signal descriptor\n");
+		return rc;
+	}
+	rc = pprocess_signal_tdesc_init(ct);
+	if (unlikely(rc)) {
+		uk_pr_err("Could not initialize signal descriptor\n");
+		return rc;
+	}
+
+	/* If CLONE_THREAD was passed, the child is assigned to the
+	 * calling process, so no further action is required.
+	 */
+	if (cl_args->flags & CLONE_THREAD)
+		return 0;
+
+	/* Initialize the new process' signal descriptor */
+	cp->signal = uk_malloc(cp->_a, sizeof(struct uk_signal_pdesc));
+	if (unlikely(!cp->signal)) {
+		uk_pr_err("Could not allocate memory\n");
+		return -ENOMEM;
+	}
+
+	/* CLONE_CLEAR_SIGHAND: Reset child's signal dispositions to default. */
+	if (cl_args->flags & CLONE_CLEAR_SIGHAND) {
+		uk_pr_debug("CLONE_SIGHAND: pid %d gets default signal dispositions\n",
+			    cpid);
+
+		if (!(cl_args->flags & CLONE_THREAD)) {
+			rc = pprocess_signal_sigaction_new(cp->_a, cp->signal);
+			if (unlikely(rc))
```

Comment:
there seem to be some errors returned throughout this function without having freed `cp->signal`.

Reply:
This is the kind of thing that should had been caught by static analysis, yet codacity is lazy on reviewing and only complains about stupid things like reducing variable scope.